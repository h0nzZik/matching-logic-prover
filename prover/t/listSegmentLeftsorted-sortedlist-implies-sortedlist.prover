claim \implies( \and( listSegmentLeftSorted( variable("H") { ArrayIntInt } 
                                    , variable("X") { Int } 
                                    , variable("Y") { Int } 
                                    , variable("F") { SetInt }
                                    , variable("MIN") { Int }
                                    , variable("MAX") { Int } 
                                    )
                    , listSorted( variable("H") { ArrayIntInt }
                                , variable("Y") { Int }
                                , variable("G") { SetInt }
                                , variable("MIN2") { Int }) 
                    , \equals( variable("K") { SetInt }, union(variable("F") { SetInt }, variable("G") { SetInt })) 
                    , disjoint(variable("F") { SetInt }, variable("G") { SetInt }) 
                    , \not(gt(variable("MAX") { Int }, variable("MIN2") { Int }))
                     )
              , \and( listSorted( variable("H") { ArrayIntInt }
                                , variable("X") { Int }
                                , variable("K") { SetInt }
                                , variable("MIN") { Int }
                                )
                     )
              )

strategy kt ; normalize
            ; ( ( simplify
                ; left-unfold-Nth(0) ;
                  ( ( right-unfold-Nth(0, 0) ; smt)
                  | ( right-unfold-Nth(0, 1) ; smt)
                  )
                )
              | ( right-unfold-Nth(0, 1)
                ; instantiate-existentials
                ; instantiate-aux
                ; instantiate-aux
                ; smt
                )
              )


// strategy search-bound(4)
