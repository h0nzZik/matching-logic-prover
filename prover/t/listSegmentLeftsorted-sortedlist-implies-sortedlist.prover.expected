<prover>
  <k>
    .
  </k>
  <goals>
    <goal>
      <id>
        root
      </id>
      <active>
        true
      </active>
      <parent>
        .
      </parent>
      <claim>
        \implies ( \and ( listSegmentLeftSorted ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , variable ( "MIN" ) { Int } , variable ( "MAX" ) { Int } , .Patterns ) , listSorted ( variable ( "H" ) { ArrayIntInt } , variable ( "Y" ) { Int } , variable ( "G" ) { SetInt } , variable ( "MIN2" ) { Int } , .Patterns ) , \equals ( variable ( "K" ) { SetInt } , union ( variable ( "F" ) { SetInt } , variable ( "G" ) { SetInt } , .Patterns ) ) , disjoint ( variable ( "F" ) { SetInt } , variable ( "G" ) { SetInt } , .Patterns ) , \not ( gt ( variable ( "MAX" ) { Int } , variable ( "MIN2" ) { Int } , .Patterns ) ) , .Patterns ) , \exists { .Patterns } \and ( listSorted ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "K" ) { SetInt } , variable ( "MIN" ) { Int } , .Patterns ) , .Patterns ) )
      </claim>
      <strategy>
        success
      </strategy>
      <trace>
        .
      </trace>
    </goal>
  </goals>
  <declarations>
    <declaration>
      axiom \forall { variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "PREV_VAL" ) { Int } , .Patterns } \iff-lfp ( listSorted ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "F" ) { SetInt } , variable ( "PREV_VAL" ) { Int } , .Patterns ) , \or ( \exists { .Patterns } \and ( \equals ( variable ( "X" ) { Int } , 0 ) , \equals ( variable ( "F" ) { SetInt } , emptyset ) , .Patterns ) , \exists { variable ( "F_I" ) { SetInt } , variable ( "VAL_I" ) { Int } , variable ( "X_I" ) { Int } , .Patterns } \and ( listSorted ( variable ( "H" ) { ArrayIntInt } , variable ( "X_I" ) { Int } , variable ( "F_I" ) { SetInt } , variable ( "VAL_I" ) { Int } , .Patterns ) , gt ( variable ( "X" ) { Int } , 0 , .Patterns ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , .Patterns ) , variable ( "X_I" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , union ( variable ( "F_I" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) ) , disjoint ( variable ( "F_I" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) , \equals ( variable ( "VAL_I" ) { Int } , select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 1 , .Patterns ) , .Patterns ) ) , gt ( variable ( "VAL_I" ) { Int } , variable ( "PREV_VAL" ) { Int } , .Patterns ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      axiom \forall { variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , variable ( "PREV_VAL" ) { Int } , variable ( "MAX" ) { Int } , .Patterns } \iff-lfp ( listSegmentLeftSorted ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , variable ( "Y" ) { Int } , variable ( "F" ) { SetInt } , variable ( "PREV_VAL" ) { Int } , variable ( "MAX" ) { Int } , .Patterns ) , \or ( \exists { .Patterns } \and ( \equals ( variable ( "X" ) { Int } , variable ( "Y" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , emptyset ) , \not ( gt ( variable ( "PREV_VAL" ) { Int } , variable ( "MAX" ) { Int } , .Patterns ) ) , .Patterns ) , \exists { variable ( "F_I" ) { SetInt } , variable ( "VAL_I" ) { Int } , variable ( "X_I" ) { Int } , .Patterns } \and ( listSegmentLeftSorted ( variable ( "H" ) { ArrayIntInt } , variable ( "X_I" ) { Int } , variable ( "Y" ) { Int } , variable ( "F_I" ) { SetInt } , variable ( "VAL_I" ) { Int } , variable ( "MAX" ) { Int } , .Patterns ) , gt ( variable ( "X" ) { Int } , 0 , .Patterns ) , \equals ( select ( variable ( "H" ) { ArrayIntInt } , variable ( "X" ) { Int } , .Patterns ) , variable ( "X_I" ) { Int } ) , \equals ( variable ( "F" ) { SetInt } , union ( variable ( "F_I" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) ) , disjoint ( variable ( "F_I" ) { SetInt } , singleton ( variable ( "X" ) { Int } , .Patterns ) , .Patterns ) , \equals ( variable ( "VAL_I" ) { Int } , select ( variable ( "H" ) { ArrayIntInt } , plus ( variable ( "X" ) { Int } , 1 , .Patterns ) , .Patterns ) ) , gt ( variable ( "VAL_I" ) { Int } , variable ( "PREV_VAL" ) { Int } , .Patterns ) , \not ( gt ( variable ( "VAL_I" ) { Int } , variable ( "MAX" ) { Int } , .Patterns ) ) , .Patterns ) , .Patterns ) )
    </declaration> <declaration>
      symbol listSegmentLeftSorted ( ArrayIntInt , Int , Int , SetInt , Int , Int , .Sorts ) : Bool
    </declaration> <declaration>
      symbol listSorted ( ArrayIntInt , Int , SetInt , Int , .Sorts ) : Bool
    </declaration>
  </declarations>
</prover>
